WooCommerce/Classes/ViewRelated/CardPresentPayments/CardReaderConnectionController.swift

```diff
@@ -0,0 +1,237 @@
+        ///
+        case idle
+
+        /// Initializing (fetching the list of any known readers)
+        ///
+        case initializing
+
+        /// Begin search for card readers
+        ///
+        case beginSearch
+
+        /// Searching for a card reader
+        ///
+        case searching
+
+        /// Ending search for card readers
+        ///
+        case endSearch
+
+        /// Found a card reader
+        ///
+        case foundReader
+
+        /// Attempting to connect to a card reader
+        ///
+        case connectToReader
+
+        /// Connected successfully to a card reader
+        ///
+        case connectedToReader
+
+        /// A failure occurred during search or connection
+        ///
+        case failed(Error)
+    }
+
+    private var state: ControllerState {
+        didSet {
+            didSetState()
+        }
+    }
+    private var fromController: UIViewController
+    private var siteID: Int64
+    private var knownCardReadersProvider: CardReaderSettingsKnownReadersProvider
+
+    private var foundReader: CardReader?
+    private var knownReaderIDs: [String]
+    private var subscriptions = Set<AnyCancellable>()
+
+    private lazy var alerts: CardReaderSettingsAlerts = {
+        CardReaderSettingsAlerts()
+    }()
+
+    init(
+        from: UIViewController,
+        forSiteID: Int64,
+        knownReadersProvider: CardReaderSettingsKnownReadersProvider
+    ) {
+        state = .idle
+        fromController = from
+        siteID = forSiteID
+        knownCardReadersProvider = knownReadersProvider
+        knownReaderIDs = []
+    }
+
+    deinit {
+        subscriptions.removeAll()
+    }
+
+    func start() {
+        self.state = .initializing
+
+        /// Subscribe to the list of known readers
+        ///
+        knownCardReadersProvider.knownReaders.sink(receiveValue: { [weak self] readerIDs in
+            self?.knownReaderIDs = readerIDs
+            /// We are now ready to actually begin the search
+            self?.state = .beginSearch
+        }).store(in: &subscriptions)
+
+    }
+}
+
+private extension CardReaderConnectionController {
+    func didSetState() {
+        switch state {
+        case .beginSearch:
+            onBeginSearch()
+        case .searching:
+            onSearching()
+        case .endSearch:
+            onEndSearch()
+        case .foundReader:
+            onFoundReader()
+        case .connectToReader:
+            onConnectToReader()
+        case .failed(let error):
+            onFailed(error: error)
+        default:
+            dismissAnyModal()
+        }
+    }
+
+    /// Begin the search for a card reader
+    ///
+    func onBeginSearch() {
+        self.state = .searching
+        let action = CardPresentPaymentAction.startCardReaderDiscovery(
+            siteID: siteID,
+            onReaderDiscovered: { [weak self] cardReaders in
+                /// Surprisingly, onReaderDiscovered may be called with an empty array
```

Comment:
I would suggest guarding here against self being nil. That would remove all the optional self in the rest of the inplementation

Reply:
Fixed in https://github.com/woocommerce/woocommerce-ios/pull/4721/commits/af4232cacb613e4b77f3cf66ea0d88a4f88aa101